skops/hub_utils/_hf_hub.py

```diff
@@ -56,7 +64,76 @@ def _validate_folder(path: Union[str, Path]):
         raise TypeError(f"Model file {model_path} does not exist.")
 
 
-def _create_config(*, model_path: str, requirements: List[str], dst: str):
+def _get_example_input(data):
+    """Returns the example input of a model.
+
+    The input is converted into a dictionary which is then stored in the config
+    file.
+
+    Parameters
+    ----------
+    data: array-like
+        The input needs to be either a ``pandas.DataFrame`` or a
+        ``numpy.ndarray``. The first 3 rows are used as example input.
+
+    Returns
+    -------
+    example_input: dict of lists
+        The example input of the model as accepted by HuggingFace's backend.
+    """
+    try:
+        import pandas as pd
+
+        if isinstance(data, pd.DataFrame):
+            return {x: data[x][:3].to_list() for x in data.columns}
+    except ImportError:
+        # pandas is not installed, the data cannot be a pandas DataFrame
+        pass
+
+    # here we convert the first three rows of the numpy array to a dict of lists
+    # to be stored in the config file
+    if isinstance(data, np.ndarray):
+        return {x: data[:3, x].tolist() for x in range(data.shape[1])}
+
+    raise ValueError("The data is not a pandas.DataFrame or a numpy.ndarray.")
+
+
+def _get_column_dtypes(data):
+    """Returns the dtype of the columns of the data.
+
+    If data is a numpy.ndarray, column names are assumed to be ``X0`` to
+    ``Xn-1``, where ``n`` is the number of columns.
+
+    Parameters
+    ----------
+    data: pandas.DataFrame or numpy.ndarray
+        The data whose columns along with their dtypes are to be returned.
+
+    Returns
+    -------
+    columns: list of tuples
+        A list of tuples of the form (column name, dtype).
+    """
+    try:
+        import pandas as pd
+
+        if isinstance(data, pd.DataFrame):
+            return list(zip(data.dtypes.index, data.dtypes.astype("str")))
+    except ImportError:
+        # pandas is not installed, the data cannot be a pandas DataFrame
+        pass
+
+    # TODO: this is going to fail for Structured Arrays. We can add support for
+    # them later if we see need for it.
+    if isinstance(data, np.ndarray):
+        return [(f"X{x}", str(data.dtype)) for x in range(data.shape[1])]
+
+    raise ValueError("The data is not a pandas.DataFrame or a numpy.ndarray.")
+
+
+def _create_config(
+    *, model_path: str, requirements: List[str], dst: str, task: str, data
```

Comment:
Yes, I agree that especially different input data types can be tricky with mypy (especially especially since numpy support is only meh). If it's possible to comfortably use mypy without enforcing it everywhere, I would be in favor (I have only used it in strict mode in the past, so no experience with non-strict).

> They also help with IDE hints when users are coding and using the API, it's not just about static type checkers.

That is, if they are indeed correct, which is currently not enforced.